platform/linux-generic/pktio/socket_xdp.c

```diff
@@ -0,0 +1,591 @@
+{
+	return _odp_promisc_mode_get_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+					pktio_entry->s.name);
+}
+
+static int sock_xdp_mac_addr_get(pktio_entry_t *pktio_entry ODP_UNUSED, void *mac_addr)
+{
+	return _odp_mac_addr_get_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				    pktio_entry->s.name, mac_addr) ? -1 : ETH_ALEN;
+}
+
+static int sock_xdp_link_status(pktio_entry_t *pktio_entry)
+{
+	return _odp_link_status_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				   pktio_entry->s.name);
+}
+
+static int sock_xdp_link_info(pktio_entry_t *pktio_entry, odp_pktio_link_info_t *info)
+{
+	return _odp_link_info_fd(pkt_priv(pktio_entry)->sock_info.helper_sock,
+				 pktio_entry->s.name, info);
+}
+
+static int sock_xdp_capability(pktio_entry_t *pktio_entry, odp_pktio_capability_t *capa)
+{
+	pkt_xdp_t *priv = pkt_priv(pktio_entry);
+
+	capa->max_input_queues  = 1U;
+	capa->max_output_queues = 1U;
+	capa->set_op.op.promisc_mode = 1U;
+	capa->set_op.op.maxlen = 1U;
+
+	capa->maxlen.equal = true;
+	capa->maxlen.min_input = _ODP_SOCKET_MTU_MIN;
+	capa->maxlen.max_input = priv->sock_info.max_mtu;
+	capa->maxlen.min_output = _ODP_SOCKET_MTU_MIN;
+	capa->maxlen.max_output = priv->sock_info.max_mtu;
+
+	capa->config.parser.layer = ODP_PROTO_LAYER_ALL;
+
+	capa->stats.pktio.all_counters = 0U;
+	capa->stats.pktin_queue.all_counters = 0U;
+	capa->stats.pktout_queue.all_counters = 0U;
+
+	return 0;
+}
+
+const pktio_if_ops_t _odp_sock_xdp_pktio_ops = {
+	/* TODO: at least stats */
+	.name = "socket_xdp",
+	.print = NULL,
+	.init_global = sock_xdp_init_global,
+	.init_local = NULL,
+	.term = NULL,
+	.open = sock_xdp_open,
+	.close = sock_xdp_close,
+	.start = NULL,
+	.stop = NULL,
+	.stats = NULL,
+	.stats_reset = NULL,
+	.pktin_queue_stats = NULL,
+	.pktout_queue_stats = NULL,
+	.extra_stat_info = NULL,
+	.extra_stats = NULL,
+	.extra_stat_counter = NULL,
+	.pktio_ts_res = NULL,
+	.pktio_ts_from_ns = NULL,
+	.pktio_time = NULL,
+	.recv = sock_xdp_recv,
+	.recv_tmo = NULL,
+	.recv_mq_tmo = NULL,
+	.fd_set = NULL,
+	.send = sock_xdp_send,
+	.maxlen_get = sock_xdp_mtu_get,
+	.maxlen_set = sock_xdp_mtu_set,
+	.promisc_mode_set = sock_xdp_promisc_mode_set,
+	.promisc_mode_get = sock_xdp_promisc_mode_get,
+	.mac_get = sock_xdp_mac_addr_get,
+	.mac_set = NULL,
+	.link_status = sock_xdp_link_status,
+	.link_info = sock_xdp_link_info,
+	.capability = sock_xdp_capability,
+	.config = NULL,
+	.input_queues_config = NULL,
+	.output_queues_config = NULL
+};
+
+static odp_bool_t sock_xdp_is_mem_src_active(void)
+{
+	return !disable_pktio;
+}
+
+static void sock_xdp_force_mem_src_disable(void)
+{
+	disable_pktio = true;
+}
+
+static void sock_xdp_adjust_block_size(uint8_t *data ODP_UNUSED, uint32_t *block_size,
+				       uint32_t *block_offset ODP_UNUSED, uint32_t *flags)
+{
+	const uint32_t size = *block_size + XDP_PACKET_HEADROOM;
+	const uint64_t ps = odp_sys_page_size();
+	/* AF_XDP requires frames to be between 2kB and page size, so with
```

Comment:
Is it system page size or page size of the pool (could use huge pages)?

Reply:
The system page size.

Reply:
OK, that's a bit problematic, since our packet pool capa is currently promising that ~60kB segments are supported.

Reply:
Indeed. Limitation is in the kernel side.